### DISCUSSION TITLE
Weird pendulum behavior in Brax Basics notebook

### DISCUSSION THREAD

**Discussion by karanprime** [google/brax#532] [Q&A] [answered]
I apologize if I am missing something fundamental. I am trying to use Jax for rigid body dynamics. As a test case, I tried manipulating the pendulum in the Brax Basics notebooks (https://colab.research.google.com/github/google/brax/blob/main/notebooks/basics.ipynb). Whatever settings I use, the triple pendulum seems to reflect from the original position (theta=0) and never crosses to the other side as expected from pendulums. I increased the number of steps (to increase the total time) and played with different initial conditions (positive and negative angles). However I cannot get it to move beyond vertical position. I have attached screenshots here:
<img width="625" alt="Screenshot 2024-10-08 at 4 22 44 PM" src="https://github.com/user-attachments/assets/1ba2c016-c538-41ab-8680-b7e95182b98e">
<img width="585" alt="Screenshot 2024-10-08 at 4 23 30 PM" src="https://github.com/user-attachments/assets/b49d68fe-87ac-4a42-bdff-c077e6c15875">

I would really appreciate any help regarding this issue. How do I make the pendulum behave better?

**Comment by AumGupta** [comment, marked as answer]
Hey @karanprime,

As far as I understood, you are trying to achieve something like the **image** below:
> <img src="https://github.com/user-attachments/assets/3b0e8b55-e848-4ff5-b338-102ce0983ec0" alt="Result" width="350"/>

## Solution:
 To achieve this, you just have to play with the **parameter** of range function used in the loop at the *end* of the given [code](https://colab.research.google.com/github/google/brax/blob/main/notebooks/basics.ipynb#scrollTo=Rfzr1eD44AUx&line=31&uniqifier=1). 

The existing loop has `100 * substeps` as parameter to `range()`:
```python
for i in range(100 * substeps): #Change 100 to larger values
  if i % substeps == 0:
    visualize(ax, state.x.pos, i / (100 * substeps)) #Change 100 to larger values
  state = jax.jit(pipeline.step)(pendulum, state, None)
```
Update it to `200 * substeps`: (For Instance)
```python
for i in range(200 * substeps): 
  if i % substeps == 0:
    visualize(ax, state.x.pos, i / (200 * substeps)) 
  state = jax.jit(pipeline.step)(pendulum, state, None)
```

<details>
<summary>Click to see exact code to reproduce my result</summary>

```python
#@title { run: "auto"}

from brax.generalized import pipeline as generalized_pipeline 
from brax.positional import pipeline as positional_pipeline 
from brax.spring import pipeline as spring_pipeline 

initial_angle = 60 #@param { type:"slider", min:0, max:90, step: 1 }
pipeline = 'generalized'  #@param ["generalized", "positional", "spring"]
step_size = "5 ms" #@param [".4 ms", "1 ms", "5 ms"]

substeps = {'.4 ms': 25, '1 ms': 10, '5 ms': 2}[step_size]
stiffness = {'.4 ms': 20000, '1 ms': 10000, '5 ms': 500}[step_size]
pipeline = {'generalized': generalized_pipeline,
            'positional': positional_pipeline,
            'spring': spring_pipeline}[pipeline]

# find reasonable stiffness and damping for the given dt
link = pendulum.link.replace(
    constraint_stiffness=jp.repeat(stiffness, 3),
    constraint_vel_damping=jp.zeros((3,)))
pendulum = pendulum.replace(link=link)
pendulum = pendulum.tree_replace({'opt.timestep': 0.01 / substeps})

init_q = jp.array([initial_angle * jp.pi / 180, 0, 0])
state = jax.jit(pipeline.init)(pendulum, init_q, jp.zeros(pendulum.qd_size()))

_, ax = plt.subplots()
plt.xlim([-3, 3])
plt.ylim([-4, 0])

for i in range(200 * substeps):
  if i % substeps == 0:
    visualize(ax, state.x.pos, i / (200 * substeps))
  state = jax.jit(pipeline.step)(pendulum, state, None)

plt.title('pendulum in motion')
plt.show()
```

These are the exact parameters I used:
<img src="https://github.com/user-attachments/assets/f3533b0a-2f95-4152-95c4-0dae6507e00a" alt="Parameters" width="500"/>
</details>

***

I hope it helped ❤️. Consider marking this as *answered* if it did.
